plugins/inputs/knx_listener/knx_listener.go

```diff
@@ -0,0 +1,190 @@
+	Addresses []string
+}
+
+type KNXListener struct {
+	ServiceType    string          `toml:"service_type"`
+	ServiceAddress string          `toml:"service_address"`
+	Measurements   []Measurement   `toml:"measurement"`
+	Log            telegraf.Logger `toml:"-"`
+
+	client      KNXInterface
+	gaTargetMap map[string]addressTarget
+	gaLogbook   map[string]bool
+
+	acc telegraf.Accumulator
+}
+
+func (kl *KNXListener) Description() string {
+	return "Listener capable of handling KNX bus messages provided through a KNX-IP Interface."
+}
+
+func (kl *KNXListener) SampleConfig() string {
+	return `
+  ## Type of KNX-IP interface.
+  ## Can be either "tunnel" or "router".
+  # service_type = "tunnel"
+
+  ## Address of the KNX-IP interface.
+  service_address = "localhost:3671"
+
+  ## Measurement definition(s)
+  # [[inputs.KNXListener.measurement]]
+  #   ## Name of the measurement
+  #   name = "temperature"
+  #   ## Datapoint-Type (DPT) of the KNX messages
+  #   dpt = "9.001"
+  #   ## List of Group-Addresses (GAs) assigned to the measurement
+  #   addresses = ["5/5/1"]
+
+  # [[inputs.KNXListener.measurement]]
+  #   name = "illumination"
+  #   dpt = "9.004"
+  #   addresses = ["5/5/3"]
+`
+}
+
+func (kl *KNXListener) Gather(_ telegraf.Accumulator) error {
+	return nil
+}
+
+func (kl *KNXListener) Start(acc telegraf.Accumulator) error {
+	// Store the accumulator for later use
+	kl.acc = acc
+
+	// Setup a logbook to track unknown GAs to avoid log-spamming
+	kl.gaLogbook = make(map[string]bool)
+
+	// Construct the mapping of Group-addresses (GAs) to DPTs and the name
+	// of the measurement
+	kl.gaTargetMap = make(map[string]addressTarget)
+	for _, m := range kl.Measurements {
+		kl.Log.Debugf("Group-address mapping for measurement %q:", m.Name)
+		for _, ga := range m.Addresses {
+			kl.Log.Debugf("  %s --> %s", ga, m.Dpt)
+			if _, ok := kl.gaTargetMap[ga]; ok {
+				return fmt.Errorf("duplicate specification of address %q", ga)
+			}
+			d, ok := dpt.Produce(m.Dpt)
+			if !ok {
+				return fmt.Errorf("cannot create datapoint-type %q for address %q", m.Dpt, ga)
+			}
+			kl.gaTargetMap[ga] = addressTarget{m.Name, d}
+		}
+	}
+
+	// Connect to the KNX-IP interface
+	kl.Log.Infof("Trying to connect to %q at %q", kl.ServiceType, kl.ServiceAddress)
+	switch kl.ServiceType {
+	case "tunnel":
+		c, err := knx.NewGroupTunnel(kl.ServiceAddress, knx.DefaultTunnelConfig)
+		if err != nil {
+			return err
+		}
+		kl.client = &c
+	case "router":
+		c, err := knx.NewGroupRouter(kl.ServiceAddress, knx.DefaultRouterConfig)
+		if err != nil {
+			return err
+		}
+		kl.client = &c
+	case "dummy":
+		c, err := NewDummyInterface()
+		if err != nil {
+			return err
+		}
+		kl.client = &c
+	default:
+		return fmt.Errorf("invalid interface type: %s", kl.ServiceAddress)
+	}
+	kl.Log.Infof("Connected!")
+
+	// Listen to the KNX bus
+	go kl.listen()
+
+	return nil
+}
+
+func (kl *KNXListener) Stop() {
+	if kl.client != nil {
+		kl.client.Close()
+	}
+}
```

Comment:
should there perhaps be a waitGroup used here and wait() called in the Stop() method so that listen() can be allowed to finish what it's doing if it's in the middle of something?